### PR TITLE
Support SQL User Management

### PR DESCRIPTION
**Pull request by FingerLeader**
Implementing list, create, and delete functionalities. You can start trying them out now.

### CODE REVIEW

File: internal/flag/flag.go [hunk 1]
```diff
@@ -70,6 +70,8 @@ const (
 	Encryption            string = "encryption"
 	Compression           string = "compression"
 	SourceType            string = "source-type"
+	UserName			  string = "name"
+	UserRole			  string = "role"
```
Comment (zhangyangyu) [inline internal/flag/flag.go:74]: ident

File: internal/cli/serverless/sqluser/sqluser.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+import (
+	"tidbcloud-cli/internal"
+
+	"github.com/spf13/cobra"
+)
+
+func SQLUserCmd(h *internal.Helper) *cobra.Command {
+	var SQLUserCmd = &cobra.Command{
+		Use:   "sql-user",
+		Short: "Manage SQL users",
```
Comment (zhangyangyu) [inline internal/cli/serverless/sqluser/sqluser.go:26]:
```suggestion
		Short: "Manage cluster SQL users",
```

File: internal/cli/serverless/sqluser/create.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+	opts := CreateOpts{
+		interactive: true,
+	}
+
+	var CreateCmd = &cobra.Command{
+		Use:         "create",
+		Short:       "Create a SQL users",
+		Aliases:     []string{"c"},
+		Annotations: make(map[string]string),
+		Example: fmt.Sprintf(`  Create a TiDB Serverless SQL user in interactive mode:
```
Comment (zhangyangyu) [inline internal/cli/serverless/sqluser/create.go:83]:
```suggestion
		Example: fmt.Sprintf(`  Create a SQL user in interactive mode:
```

File: internal/cli/serverless/sqluser/create.go [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+}
+
+func CreateCmd(h *internal.Helper) *cobra.Command {
+	opts := CreateOpts{
+		interactive: true,
+	}
+
+	var CreateCmd = &cobra.Command{
+		Use:         "create",
+		Short:       "Create a SQL users",
```
Comment (zhangyangyu) [inline internal/cli/serverless/sqluser/create.go:80]:
```suggestion
		Short:       "Create a SQL user",
```

File: internal/cli/serverless/sqluser/create.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+	CreateCmd.Flags().StringP(flag.Output, flag.OutputShort, output.HumanFormat, flag.OutputHelp)
+	CreateCmd.Flags().StringP(flag.ClusterID, flag.ClusterIDShort, "", "The ID of the cluster.")
+	CreateCmd.Flags().StringP(flag.UserName, "", "", "The name of the SQL user.")
+	CreateCmd.Flags().StringP(flag.Password, "", "", "The password of the SQL user.")
+	CreateCmd.Flags().StringP(flag.UserRole, "", "", "The built-in role of the SQL user.")
+
+	return CreateCmd
+}
+
+func GetBuiltInRole(userRole string) string {
```
Comment (zhangyangyu) [inline internal/cli/serverless/sqluser/create.go:222]:
```suggestion
func GetBuiltinRole(userRole string) string {
```

File: internal/cli/serverless/sqluser/create.go [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+	role := strings.ToLower(userRole)
+	switch role {
+	case util.ADMIN:
+		role = util.ADMIN_ROLE
+	case util.READWRITE:
+		role = util.READWRITE_ROLE
+	case util.READONLY:
+		role = util.READONLY_ROLE
+	default:
+		role = userRole
```
Comment (zhangyangyu) [inline internal/cli/serverless/sqluser/create.go:232]: if the input is not the three, won't the API return error? it's not a valid builtin role right？

File: internal/cli/serverless/sqluser/delete.go [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+					if err == terminal.InterruptErr {
+						return util.InterruptError
+					} else {
+						return err
+					}
+				}
+
+				if userInput != confirmed {
+					return errors.New("incorrect confirm string entered, skipping branch deletion")
+				}
```
Comment (shiyuhang0) [inline internal/cli/serverless/sqluser/delete.go:157]: typo. branch deletion

File: internal/cli/serverless/sqluser/create.go [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+			_, err = fmt.Fprintln(h.IOStreams.Out, color.GreenString("SQL user %s is created", userName))
+			if err != nil {
+				return err
+			}
+			return nil
+
+		},
+	}
+
+	CreateCmd.Flags().StringP(flag.Output, flag.OutputShort, output.HumanFormat, flag.OutputHelp)
```
Comment (shiyuhang0) [inline internal/cli/serverless/sqluser/create.go:213]: seems CreateCmd does not need output flag

File: internal/flag/flag.go [hunk 9]
```diff
@@ -70,6 +70,8 @@ const (
 	Encryption            string = "encryption"
 	Compression           string = "compression"
 	SourceType            string = "source-type"
+	UserName			  string = "name"
```
Comment (shiyuhang0) [inline internal/flag/flag.go:73]: user or user-name should be better than `name`

**Comment by FingerLeader** [comment]
@zhangyangyu @shiyuhang0 conflicts resolved

File: internal/cli/serverless/cluster.go [hunk 10]
```diff
@@ -43,5 +44,7 @@ func Cmd(h *internal.Helper) *cobra.Command {
 	serverlessCmd.AddCommand(export.Cmd(h))
 	serverlessCmd.AddCommand(SpendingLimitCmd(h))
 	serverlessCmd.AddCommand(RegionCmd(h))
+
+	serverlessCmd.AddCommand(sqluser.SQLUserCmd(h))
```
Comment (zhangyangyu) [inline internal/cli/serverless/cluster.go:48]: adjust the order as we discussed. I think after `shell` before `import`.

File: internal/cli/serverless/sqluser/create.go [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+					return errors.Trace(err)
+				}
+
+				userRole, err = cloud.GetSelectedBuiltinRole()
+				if err != nil {
+					return err
+				}
+
+				// variables for input
+				fmt.Fprintln(h.IOStreams.Out, color.BlueString("Please input the following options"))
```
Comment (zhangyangyu) [inline internal/cli/serverless/sqluser/create.go:152]: seems other command don't have this prompt

File: internal/flag/flag.go [hunk 12]
```diff
@@ -70,6 +70,8 @@ const (
 	Encryption            string = "encryption"
 	Compression           string = "compression"
 	SourceType            string = "source-type"
+	UserName              string = "user"
```
Comment (zhangyangyu) [inline internal/flag/flag.go:73]: already have this flag

File: internal/util/search.go [hunk 13]
```diff
@@ -41,3 +41,39 @@ func IsUnderTiUP(binpath string) bool {
...
+		case ADMIN_ROLE, ADMIN:
+			displayRole = ADMIN_DISPLAY
+		case READWRITE_ROLE, READWRITE:
+			displayRole = READWRITE_DISPLAY
+		case READONLY_ROLE, READONLY:
+			displayRole = READONLY_DISPLAY
+		default:
+			displayRole = role
+		}
+
```
Comment (zhangyangyu) [inline internal/util/search.go:63]: I'm a bit lost why we need to apply this logic to every role? Isn't this specific to builtinrole?

File: internal/cli/serverless/sqluser/create.go [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+		return "", errors.Trace(err)
+	}
+
+	return cluster.Payload.UserPrefix, nil
+}
+
+func GetBuiltinRoleAndCustomRoles(userRole string) (string, []string) {
+	lowerRole := strings.ToLower(userRole)
+	switch lowerRole {
+	case util.ADMIN:
```
Comment (zhangyangyu) [inline internal/cli/serverless/sqluser/create.go:292]: when in interactive mode, shouldn't this be `util.ADMIN_DISPLAY`?

File: internal/cli/serverless/sqluser/create.go [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+
+	var CreateCmd = &cobra.Command{
+		Use:         "create",
+		Short:       "Create a SQL user",
+		Aliases:     []string{"c"},
+		Annotations: make(map[string]string),
+		Example: fmt.Sprintf(`  Create a SQL user in interactive mode:
+$ %[1]s serverless sql-user create
+
+Create a TiDB Serverless SQL user in non-interactive mode:
```
Comment (zhangyangyu) [inline internal/cli/serverless/sqluser/create.go:87]:
```suggestion
Create a SQL user in non-interactive mode:
```

File: internal/cli/serverless/sqluser/create.go [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+					return err
+				}
+				projectID := project.ID
+
+				cluster, err := cloud.GetSelectedCluster(ctx, projectID, h.QueryPageSize, d)
+				if err != nil {
+					return err
+				}
+				clusterID = cluster.ID
+				userPrefix, err = GetUserPrefix(ctx, d, clusterID)
```
Comment (zhangyangyu) [inline internal/cli/serverless/sqluser/create.go:141]: why call `GetUserPrefix` here? we could already get userprefix from cluster object

File: internal/cli/serverless/sqluser/create.go [hunk 17]
```diff
@@ -0,0 +1,302 @@
...
+
+	CreateCmd.Flags().StringP(flag.ClusterID, flag.ClusterIDShort, "", "The ID of the cluster.")
+	CreateCmd.Flags().StringP(flag.UserName, "", "", "The name of the SQL user.")
+	CreateCmd.Flags().StringP(flag.Password, "", "", "The password of the SQL user.")
+	CreateCmd.Flags().StringP(flag.UserRole, "", "", "The built-in role of the SQL user.")
+
+	return CreateCmd
+}
+
+// func GetBuiltinRole(userRole string) string {
```
Comment (zhangyangyu) [inline internal/cli/serverless/sqluser/create.go:234]: delete these

File: internal/cli/serverless/sqluser/create.go [hunk 18]
```diff
@@ -0,0 +1,302 @@
...
+			t.Placeholder = "Password"
+			t.EchoMode = textinput.EchoPassword
+			t.EchoCharacter = '•'
+		}
+		m.Inputs[v] = t
+	}
+	return m
+}
+
+func GetUserPrefix(ctx context.Context, d cloud.TiDBCloudClient, clusterID string) (string, error) {
```
Comment (zhangyangyu) [inline internal/cli/serverless/sqluser/create.go:278]: need to be public?

File: internal/cli/serverless/sqluser/create.go [hunk 19]
```diff
@@ -0,0 +1,302 @@
...
+
+	cluster, err := d.GetCluster(params)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	return cluster.Payload.UserPrefix, nil
+}
+
+func GetBuiltinRoleAndCustomRoles(userRole string) (string, []string) {
```
Comment (zhangyangyu) [inline internal/cli/serverless/sqluser/create.go:289]: need to be public?

File: internal/util/constant.go [hunk 20]
```diff
@@ -19,3 +19,27 @@ const (
...
+	READWRITE = "readwrite"
+	READONLY  = "readonly"
+)
+
+// Server accepted built-in SQL User Role types
+const (
+	ADMIN_ROLE     = "role_admin"
+	READWRITE_ROLE = "role_readwrite"
+	READONLY_ROLE  = "role_readonly"
+)
```
[the remark this answers is not in the document]
Comment (FingerLeader) [reply internal/util/constant.go:35]: User input is `admin` for simplicity, but the API requires the parameter to be `role_admin`

File: internal/util/search.go (hunk 13 above)
Comment (FingerLeader) [reply internal/util/search.go:63]: This is a mistake, the judgement should only apply to builtinRole

File: internal/cli/serverless/sqluser/create.go (hunk 14 above)
Comment (FingerLeader) [reply internal/cli/serverless/sqluser/create.go:292]: > when in interactive mode, shouldn't this be `util.ADMIN_DISPLAY`?

The return value of this function is for API

Comment (FingerLeader) [reply internal/cli/serverless/sqluser/create.go:292]: > And what if in non-interactive mode, the user passed in `role_admin, custom1, custom2`?

Yes, and I think there will be a bug that when user inputs `admin, custom1, custom2`, the `admin` should be converted to `role_admin`

File: internal/cli/serverless/sqluser/create.go (hunk 16 above)
Comment (FingerLeader) [reply internal/cli/serverless/sqluser/create.go:141]: It's a mistake

File: internal/cli/serverless/sqluser/create.go [hunk 21]
```diff
@@ -0,0 +1,302 @@
...
+
+	for k, v := range createSQLUserField {
+		t := textinput.New()
+		t.Cursor.Style = config.CursorStyle
+		t.CharLimit = 32
+
+		switch k {
+		case flag.UserName:
+			// add a prefix showing the user prefix
+			t.Placeholder = "User Name"
```
[the remark this answers is not in the document]
Comment (FingerLeader) [reply internal/cli/serverless/sqluser/create.go:263]: I have no idea how to add this `>`, this prefix is added by using a setting of `textinput.Model`

File: internal/cli/serverless/sqluser/create.go [hunk 22]
```diff
@@ -0,0 +1,302 @@
...
+type CreateOpts struct {
+	interactive bool
+}
+
+const (
+	WaitInterval = 5 * time.Second
+	WaitTimeout  = 2 * time.Minute
+)
+
+var createSQLUserField = map[string]int{
```
Comment (xuanyu66) [inline internal/cli/serverless/sqluser/create.go:51]: Why not use enum?

File: internal/cli/serverless/sqluser/create.go [hunk 23]
```diff
@@ -0,0 +1,302 @@
...
+			}
+			return nil
+
+		},
+	}
+
+	CreateCmd.Flags().StringP(flag.ClusterID, flag.ClusterIDShort, "", "The ID of the cluster.")
+	CreateCmd.Flags().StringP(flag.UserName, "", "", "The name of the SQL user.")
+	CreateCmd.Flags().StringP(flag.Password, "", "", "The password of the SQL user.")
+	CreateCmd.Flags().StringP(flag.UserRole, "", "", "The built-in role of the SQL user.")
```
Comment (xuanyu66) [inline internal/cli/serverless/sqluser/create.go:229]: It should provide valid values for the user, like https://github.com/tidbcloud/tidbcloud-cli/blob/main/internal/cli/serverless/dataimport/start/start.go#L150

File: internal/cli/serverless/sqluser/create.go [hunk 24]
```diff
@@ -0,0 +1,302 @@
...
+func (c CreateOpts) NonInteractiveFlags() []string {
+	return []string{
+		flag.ClusterID,
+		flag.UserName,
+		flag.Password,
+		flag.UserRole,
+	}
+}
+
+func (c CreateOpts) RequiredFlags() []string {
```
Comment (xuanyu66) [inline internal/cli/serverless/sqluser/create.go:65]: It is the same as NonInteractiveFlags(), could they be merged?

File: internal/cli/serverless/sqluser/delete.go [hunk 25]
```diff
@@ -0,0 +1,180 @@
...
+				if err != nil {
+					return errors.Trace(err)
+				}
+				userName = uName
+				clusterID = cID
+			}
+
+			if !force {
+				if !h.IOStreams.CanPrompt {
+					return fmt.Errorf("the terminal doesn't support prompt, please run with --force to delete the branch")
```
Comment (xuanyu66) [inline internal/cli/serverless/sqluser/delete.go:137]: not branch

File: internal/cli/serverless/sqluser/list.go [hunk 26]
```diff
@@ -0,0 +1,166 @@
...
+
+	var listCmd = &cobra.Command{
+		Use:         "list",
+		Short:       "List all accessible SQL users",
+		Aliases:     []string{"ls"},
+		Annotations: make(map[string]string),
+		Example: fmt.Sprintf(`  List the SQL users:
+  $ %[1]s user list
+
+  Get a TiDB Serverless cluster in non-interactive mode:
```
Comment (xuanyu66) [inline internal/cli/serverless/sqluser/list.go:57]: please check the description

**Review by xuanyu66 — CHANGES_REQUESTED** [review]

File: internal/cli/serverless/sqluser/create.go (hunk 23 above)
Comment (FingerLeader) [reply internal/cli/serverless/sqluser/create.go:227]: > It should provide valid values for the user, like https://github.com/tidbcloud/tidbcloud-cli/blob/main/internal/cli/serverless/dataimport/start/start.go#L150

Do you mean the `UserRole`? Users are free to choose the content they wish to fill in

File: internal/cli/serverless/sqluser/create.go (hunk 22 above)
Comment (FingerLeader) [reply internal/cli/serverless/sqluser/create.go:51]: Using an enum is also OK, just referencing other implementations

File: internal/cli/serverless/sqluser/create.go (hunk 24 above)
Comment (FingerLeader) [reply internal/cli/serverless/sqluser/create.go:65]: Do you mean just keep one of them? I think they have different meaning

Comment (xuanyu66) [reply internal/cli/serverless/sqluser/create.go:65]: LGTM

File: internal/cli/serverless/backup/delete.go [hunk 27]
```diff
@@ -144,7 +144,7 @@ func DeleteCmd(h *internal.Helper) *cobra.Command {
 				}
 
 				if userInput != confirmed {
-					return errors.New("incorrect confirm string entered, skipping branch deletion")
+					return errors.New("incorrect confirm string entered, skipping buckup deletion")
```
Comment (xuanyu66) [inline internal/cli/serverless/backup/delete.go:147]:
```suggestion
					return errors.New("incorrect confirm string entered, skipping backup deletion")
```

File: internal/cli/serverless/sqluser/create.go [hunk 28]
```diff
@@ -0,0 +1,295 @@
...
+
+func CreateCmd(h *internal.Helper) *cobra.Command {
+	opts := CreateOpts{
+		interactive: true,
+	}
+
+	var CreateCmd = &cobra.Command{
+		Use:         "create",
+		Short:       "Create a SQL user",
+		Aliases:     []string{"c"},
```
Comment (xuanyu66) [inline internal/cli/serverless/sqluser/create.go:82]: why add this alias?

**Review by xuanyu66 — CHANGES_REQUESTED** [review]

**Review by zhangyangyu — DISMISSED** [review]

**Review by xuanyu66 — APPROVED** [review]

**Review by zhangyangyu — APPROVED** [review]

**Pull request merged**